lib/semantics/tools.cc

```diff
@@ -328,6 +337,28 @@ const Symbol *FindFunctionResult(const Symbol &symbol) {
   return nullptr;
 }
 
+const Symbol *GetAssociatedVariable(const AssocEntityDetails &details) {
+  if (const MaybeExpr & expr{details.expr()}) {
+    if (evaluate::IsVariable(*expr)) {
+      if (const Symbol * varSymbol{evaluate::GetLastSymbol(*expr)}) {
+        return GetAssociationRoot(*varSymbol);
+      }
+    }
+  }
+  return nullptr;
+}
+
+// Return the Symbol of the variable of a construct association, if it exists
+// Return nullptr if the name is associated with an expression
+const Symbol *GetAssociationRoot(const Symbol &symbol) {
+  const Symbol &ultimate{symbol.GetUltimate()};
+  if (ultimate.has<AssocEntityDetails>()) {  // construct association
+    return GetAssociatedVariable(ultimate.get<AssocEntityDetails>());
```

Comment:
This could be simplified slightly:
```
if (const auto *details{ultimate.detailsIf<AssocEntityDetails>()}) {
  return GetAssociatedVariable(*details);
```

Reply:
Thanks, I'll change this.